rfcs/013-gitlab-spam-checker/013-gitlab-spam-checker.md

```diff
@@ -0,0 +1,93 @@
+
+## Key Assumptions
+
+- The proposed machine learning model will be built in consultation with the innovation team. The machine learning model's specifics are out of this RFC's scope.
+
+## Possible Solutions
+
+Possible solutions include a service that only checks for spam at regular intervals. This solution is less than ideal because of the delay between the spam being posted and the spam being detected and removed, during which spam could already be harmful. Another solution is to monitor GitLab for spam manually. This solution is also unsuitable because it is time-consuming.
+
+GitLab has its [Spamcheck](https://docs.gitlab.com/ee/user/admin_area/reporting/spamcheck.html) service, which can be used to check for spam issues and snippets. This service does not check for spam in notes, projects, or user profiles. Currently, it is not possible to configure the service to delete spam automatically. Deploying GitLab's Spamcheck service also requires GitLab Enterprise Edition, which RenkuLab.io's GitLab is not currently running on.
+
+GitLab can make use of Akismet, a third-party-hosted AI spam filter. However, this only checks for spam on issues and notes, needs training, and would involve handing data over to a third party in the US.
+
+## Proposed Solution
+
+The proposed spam management service will integrate with GitLab via system hooks. When notes, projects, or users are created or updated, GitLab will send system hooks to a spam management service. System hooks request examples can be found [here](https://docs.gitlab.com/ee/administration/system_hooks.html#hooks-request-example).
+
+System hooks triggered on group, project, and user events can be triggered at the system level. Issues and comments at a project level would require a project-level system hook. This might be able to be configured when a project is added to KG. Project-level webhook events examples can be found [here](https://docs.gitlab.com/ee/user/project/integrations/webhook_events.html).
+
+Snippets are unable to trigger system hooks. Therefore a job triggered at regular intervals will be required to check for new snippets. This can be done by checking the snippet API for new snippets. Documentation on the snippet API can be found [here](https://docs.gitlab.com/ee/api/snippets.html).
+
+Ideally, the solution will also be able to check abuse reports manually submitted by users in GitLab and check the reported URL in the abuse report for spam content.
+
+A feedback mechanism would also be implemented, allowing the machine learning model to learn from mistakes and improve over time. This could be done by storing content in an S3 bucket and regularly re-training the model on a dataset including new data. The system should also keep a log of decisions made and the confidence level of the decision.
```

Comment:
This feels somehow like the second phase of the work; the objective in the near term would be to get a variant operational which has no specific feedback mechanisms?

Reply:
I don't think so - we need to be able to log what decisions are made and why from the start. Our current hacked together spam management scripts can do this so I see no reason why the service shouldn't be able to as well.

Reply:
I'd also lean into having the feedback mechanism from the start and I would spend some words on its design.